src/compiler/parser.ts

```diff
@@ -4517,6 +4523,12 @@ module ts {
         }
 
         function parseClassDeclaration(fullStart: number, modifiers: ModifiersArray): ClassDeclaration {
+            // In ES6 specification, All parts of a ClassDeclaration or a ClassExpression are strict mode code
+            if (languageVersion >= ScriptTarget.ES6) {

```

Comment:
I would always enter strict mode once we're in a class.  If that's ES semantics, then that's what we should follow.